mypy/checker.py

```diff
@@ -5572,6 +5568,25 @@ def visit_match_stmt(self, s: MatchStmt) -> None:
             with self.binder.frame_context(can_skip=False, fall_through=2):
                 pass
 
+    def _make_named_statement_for_match(self, s: MatchStmt) -> Expression:
+        """Construct a fake NameExpr for inference if a match clause is complex."""
+        subject = s.subject
+        if isinstance(subject, (NameExpr, AssignmentExpr)):
```

Comment:
What I wanted is something like this:
```python
        if isinstance(subject, (IndexExpr, MemberExpr, NameExpr) and literal(subject):
            <use existing expression>
        else:
            <create fake name>
```
plus an explicit comment that this should be kept in sync with corresponding logic in binder. This way it will be easier to spot a bug if something changes in binder.

Reply:
Hm, okay, you finally made me realize that the problem with walrus is not situated here - thanks! I think the new version should be closer.